[RISCV][TTI] Support cost of f16 FCmp using zvfhmin in the absence of…

arcbbb opened the pull request:
… zvfh

With zvfhmin, the cost of f16 FCmp is determined by the cost of widening two operands from f16 to f32, plus the cost of performing an f32 FCmp

jacquesguan commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -1398,12 +1398,28 @@ InstructionCost RISCVTTIImpl::getCmpSelInstrCost(unsigned Opcode, Type *ValTy,
     // InvalidCost for scalable vector.
-    if ((ValTy->getScalarSizeInBits() == 16 && !ST->hasVInstructionsF16()) ||
+    if ((ValTy->getScalarSizeInBits() == 16 && !ST->hasVInstructionsF16() &&
+         !ST->hasVInstructionsF16Minimal()) ||
```
Using `hasVInstructionsF16Minimal` is enough since zvfh implies zvfhmin in llvm now.

arcbbb replied:
Thanks for the information!

arcbbb commented:
soft ping

preames reviewed (commented):
Please precommit the extraction of rvv-fcmp-f16.ll so that the test changes are easier to follow.  You should probably also add a test line with no half support.

topperc commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -1398,12 +1398,28 @@ InstructionCost RISCVTTIImpl::getCmpSelInstrCost(unsigned Opcode, Type *ValTy,
+                                      CostKind);
+      InstructionCost CmpCost =
+          getCmpSelInstrCost(Opcode, VecF32Ty, CondTy, VecPred, CostKind, I);
+      return VecF32LT.first * WidenCost + CmpCost;
```
Should this be `VecF32LT.first * (WidenCost + CmpCost)`?

arcbbb replied:
I discovered an issue with handling `vscale x 32 x f16`, since the vtype doesn't work for vfwcvt. To fix this, I switched to using getCastInstrCost to calculate the split and widen costs.

arcbbb committed:
[RISCV][TTI] Support cost of f16 FCmp using zvfhmin in the absence of zvfh

With zvfhmin, the cost of f16 FCmp is determined by the cost of widening
two
operands from f16 to f32, plus the cost of performing an f32 FCmp

arcbbb commented:
I have precommitted the fcmp f16 test and rebased.
